Agent error while installing tools (`installed.json` corruption)

Environment details: macOS, Agent 1.6.0
Reported by @charlesbones

The agent starts and the closes immediately with a panic, with the following logs:
```
ERRO[0001] Error downloading serial-discovery: invalid character ‘y’ after top-level value
panic: invalid character ‘y’ after top-level value
goroutine 69 [running]:
main.(*SerialPortList).runSerialDiscovery(0x1400027ecc0?)
	/Users/runner/work/arduino-create-agent/arduino-create-agent/serial.go:130 +0x568
main.(*SerialPortList).Run(0x140001a0180?)
	/Users/runner/work/arduino-create-agent/arduino-create-agent/serial.go:118 +0x2c
created by main.loop in goroutine 50
	/Users/runner/work/arduino-create-agent/arduino-create-agent/main.go:410 +0x1244
```

Looking at the message, this might be a `installed.json` corruption problem. See also this previous issue: https://github.com/arduino/arduino-create-agent/pull/635

Here is the content of the JSON file on the same machine (note the wrong data at the end):
```json
{"arduino-fwuploader":"/Users/carlos/.arduino-create/arduino/arduino-fwuploader/2.4.1","arduino-fwuploader-2.4.1":"/Users/carlos/.arduino-create/arduino/arduino-fwuploader/2.4.1","bossac":"/Users/carlos/.arduino-create/arduino/bossac/1.7.0","bossac-1.7.0":"/Users/carlos/.arduino-create/arduino/bossac/1.7.0","bossac-1.9.1-arduino5":"/Users/carlos/.arduino-create/arduino/bossac/1.9.1-arduino5","fwupdater":"/Users/carlos/.arduino-create/arduino/fwupdater/0.1.12","fwupdater-0.1.12":"/Users/carlos/.arduino-create/arduino/fwupdater/0.1.12","rp2040tools":"/Users/carlos/.arduino-create/arduino/rp2040tools/1.0.6","rp2040tools-1.0.6":"/Users/carlos/.arduino-create/arduino/rp2040tools/1.0.6","serial-discovery":"/Users/carlos/.arduino-create/builtin/serial-discovery/1.4.1","serial-discovery-1.4.0":"/Users/carlos/.arduino-create/builtin/serial-discovery/1.4.0","serial-discovery-1.4.1":"/Users/carlos/.arduino-create/builtin/serial-discovery/1.4.1"}y/1.4.1"}
```

### Additional context

#### Additional reports

- https://forum.arduino.cc/t/opening-port-at-1200bps-permission-denied/1285652/7
- https://forum.arduino.cc/t/arduino-cloud-agent-error-on-macos-sonoma/1287887
- https://forum.arduino.cc/t/cloud-agent-crashes-invalid-character-y-after-top-level-value/1288214

--- a/tools/download.go
+++ b/tools/download.go
@@ -25,7 +25,6 @@ import (
 
 	"github.com/arduino/arduino-create-agent/gen/tools"
 	"github.com/arduino/arduino-create-agent/utilities"
-	"github.com/arduino/arduino-create-agent/v2/pkgs"
 )
 
 // Download will parse the index at the indexURL for the tool to download.
@@ -45,8 +44,8 @@ import (
 // if it already exists.
 func (t *Tools) Download(pack, name, version, behaviour string) error {
 
-	tool := pkgs.New(t.index, t.directory.String(), behaviour)
-	_, err := tool.Install(context.Background(), &tools.ToolPayload{Name: name, Version: version, Packager: pack})
+	t.tools.SetBehaviour(behaviour)
+	_, err := t.tools.Install(context.Background(), &tools.ToolPayload{Name: name, Version: version, Packager: pack})
 	if err != nil {
 		return err
 	}

--- a/tools/download_test.go
+++ b/tools/download_test.go
@@ -160,3 +160,23 @@ func TestDownload(t *testing.T) {
 		})
 	}
 }
+
+func TestCorruptedInstalled(t *testing.T) {
+	// prepare the test environment
+	tempDir := t.TempDir()
+	tempDirPath := paths.New(tempDir)
+	testIndex := index.Resource{
+		IndexFile:   *paths.New("testdata", "test_tool_index.json"),
+		LastRefresh: time.Now(),
+	}
+	corruptedJSON := tempDirPath.Join("installed.json")
+	fileJSON, err := corruptedJSON.Create()
+	require.NoError(t, err)
+	defer fileJSON.Close()
+	_, err = fileJSON.Write([]byte("Hello"))
+	require.NoError(t, err)
+	testTools := New(tempDirPath, &testIndex, func(msg string) { t.Log(msg) })
+	// Download the tool
+	err = testTools.Download("arduino-test", "avrdude", "6.3.0-arduino17", "keep")
+	require.NoError(t, err)
+}

--- a/tools/tools.go
+++ b/tools/tools.go
@@ -22,6 +22,7 @@ import (
 	"sync"
 
 	"github.com/arduino/arduino-create-agent/index"
+	"github.com/arduino/arduino-create-agent/v2/pkgs"
 	"github.com/arduino/go-paths-helper"
 	"github.com/xrash/smetrics"
 )
@@ -47,6 +48,7 @@ type Tools struct {
 	logger    func(msg string)
 	installed map[string]string
 	mutex     sync.RWMutex
+	tools     *pkgs.Tools
 }
 
 // New will return a Tool object, allowing the caller to execute operations on it.
@@ -60,6 +62,7 @@ func New(directory *paths.Path, index *index.Resource, logger func(msg string)) 
 		logger:    logger,
 		installed: map[string]string{},
 		mutex:     sync.RWMutex{},
+		tools:     pkgs.New(index, directory.String(), "replace"),
 	}
 	_ = t.readMap()
 	return t

--- a/v2/pkgs/tools.go
+++ b/v2/pkgs/tools.go
@@ -331,6 +331,11 @@ func (t *Tools) writeInstalled(path string) error {
 	return os.WriteFile(installedFile, data, 0644)
 }
 
+// SetBehaviour sets the download behaviour to either keep or replace
+func (t *Tools) SetBehaviour(behaviour string) {
+	t.behaviour = behaviour
+}
+
 func pathExists(path string) bool {
 	_, err := os.Stat(path)
 	if err == nil {

--- a/v2/pkgs/tools.go
+++ b/v2/pkgs/tools.go
@@ -29,6 +29,7 @@ import (
 	"path/filepath"
 	"runtime"
 	"strings"
+	"sync"
 
 	"github.com/arduino/arduino-create-agent/gen/tools"
 	"github.com/arduino/arduino-create-agent/index"
@@ -60,6 +61,7 @@ type Tools struct {
 	index     *index.Resource
 	folder    string
 	behaviour string
+	mutex     sync.RWMutex
 }
 
 // New will return a Tool object, allowing the caller to execute operations on it.
@@ -70,6 +72,7 @@ func New(index *index.Resource, folder, behaviour string) *Tools {
 		index:     index,
 		folder:    folder,
 		behaviour: behaviour,
+		mutex:     sync.RWMutex{},
 	}
 }
 
@@ -187,7 +190,7 @@ func (t *Tools) Install(ctx context.Context, payload *tools.ToolPayload) (*tools
 		}
 		if ok && pathExists(location) {
 			// overwrite the default tool with this one
-			err := writeInstalled(t.folder, path)
+			err := t.writeInstalled(path)
 			if err != nil {
 				return nil, err
 			}
@@ -245,7 +248,7 @@ func (t *Tools) install(ctx context.Context, path, url, checksum string) (*tools
 	}
 
 	// Write installed.json for retrocompatibility with v1
-	err = writeInstalled(t.folder, path)
+	err = t.writeInstalled(path)
 	if err != nil {
 		return nil, err
 	}
@@ -311,9 +314,11 @@ func checkInstalled(folder, key string) (string, bool, error) {
 	return location, ok, err
 }
 
-func writeInstalled(folder, path string) error {
+func (t *Tools) writeInstalled(path string) error {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
 	// read installed.json
-	installedFile, err := utilities.SafeJoin(folder, "installed.json")
+	installedFile, err := utilities.SafeJoin(t.folder, "installed.json")
 	if err != nil {
 		return err
 	}
@@ -325,7 +330,7 @@ func writeInstalled(folder, path string) error {
 	parts := strings.Split(path, string(filepath.Separator))
 	tool := parts[len(parts)-2]
 	toolWithVersion := fmt.Sprint(tool, "-", parts[len(parts)-1])
-	toolFile, err := utilities.SafeJoin(folder, path)
+	toolFile, err := utilities.SafeJoin(t.folder, path)
 	if err != nil {
 		return err
 	}

--- a/v2/pkgs/tools.go
+++ b/v2/pkgs/tools.go
@@ -290,8 +290,8 @@ func rename(base string) extract.Renamer {
 }
 
 func (t *Tools) readInstalled() error {
-	t.mutex.Lock()
-	defer t.mutex.Unlock()
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
 	// read installed.json
 	installedFile, err := utilities.SafeJoin(t.folder, "installed.json")
 	if err != nil {
@@ -305,8 +305,8 @@ func (t *Tools) readInstalled() error {
 }
 
 func (t *Tools) writeInstalled(path string) error {
-	t.mutex.RLock()
-	defer t.mutex.RUnlock()
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
 
 	parts := strings.Split(path, string(filepath.Separator))
 	tool := parts[len(parts)-2]

--- a/v2/pkgs/tools.go
+++ b/v2/pkgs/tools.go
@@ -61,6 +61,7 @@ type Tools struct {
 	index     *index.Resource
 	folder    string
 	behaviour string
+	installed map[string]string
 	mutex     sync.RWMutex
 }
 
@@ -68,12 +69,15 @@ type Tools struct {
 // The New function will accept an index as parameter (used to download the indexes)
 // and a folder used to download the indexes
 func New(index *index.Resource, folder, behaviour string) *Tools {
-	return &Tools{
+	t := &Tools{
 		index:     index,
 		folder:    folder,
 		behaviour: behaviour,
+		installed: map[string]string{},
 		mutex:     sync.RWMutex{},
 	}
+	t.readInstalled()
+	return t
 }
 
 // Installedhead is here only because it was required by the front-end.
@@ -184,10 +188,7 @@ func (t *Tools) Install(ctx context.Context, payload *tools.ToolPayload) (*tools
 	key := correctTool.Name + "-" + correctTool.Version
 	// Check if it already exists
 	if t.behaviour == "keep" && pathExists(t.folder) {
-		location, ok, err := checkInstalled(t.folder, key)
-		if err != nil {
-			return nil, err
-		}
+		location, ok := t.installed[key]
 		if ok && pathExists(location) {
 			// overwrite the default tool with this one
 			err := t.writeInstalled(path)
@@ -288,44 +289,24 @@ func rename(base string) extract.Renamer {
 	}
 }
 
-func readInstalled(installedFile string) (map[string]string, error) {
-	// read installed.json
-	installed := map[string]string{}
-	data, err := os.ReadFile(installedFile)
-	if err == nil {
-		err = json.Unmarshal(data, &installed)
-		if err != nil {
-			return nil, err
-		}
-	}
-	return installed, nil
-}
-
-func checkInstalled(folder, key string) (string, bool, error) {
-	installedFile, err := utilities.SafeJoin(folder, "installed.json")
-	if err != nil {
-		return "", false, err
-	}
-	installed, err := readInstalled(installedFile)
-	if err != nil {
-		return "", false, err
-	}
-	location, ok := installed[key]
-	return location, ok, err
-}
-
-func (t *Tools) writeInstalled(path string) error {
-	t.mutex.RLock()
-	defer t.mutex.RUnlock()
+func (t *Tools) readInstalled() error {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
 	// read installed.json
 	installedFile, err := utilities.SafeJoin(t.folder, "installed.json")
 	if err != nil {
 		return err
 	}
-	installed, err := readInstalled(installedFile)
+	data, err := os.ReadFile(installedFile)
 	if err != nil {
 		return err
 	}
+	return json.Unmarshal(data, &t.installed)
+}
+
+func (t *Tools) writeInstalled(path string) error {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
 
 	parts := strings.Split(path, string(filepath.Separator))
 	tool := parts[len(parts)-2]
@@ -334,10 +315,15 @@ func (t *Tools) writeInstalled(path string) error {
 	if err != nil {
 		return err
 	}
-	installed[tool] = toolFile
-	installed[toolWithVersion] = toolFile
+	t.installed[tool] = toolFile
+	t.installed[toolWithVersion] = toolFile
 
-	data, err := json.Marshal(installed)
+	data, err := json.Marshal(t.installed)
+	if err != nil {
+		return err
+	}
+
+	installedFile, err := utilities.SafeJoin(t.folder, "installed.json")
 	if err != nil {
 		return err
 	}

